modules/distribution/keeper/querier.go

```diff
@@ -0,0 +1,242 @@
+	ctx, _ = ctx.CacheContext()
+	if !k.HasValidatorDistInfo(ctx, params.ValidatorAddress) {
+		return []byte{}, types.ErrNoValidatorDistInfo(types.DefaultCodespace)
+	}
+	vdi := k.GetValidatorDistInfo(ctx, params.ValidatorAddress)
+	res, errRes := codec.MarshalJSONIndent(k.cdc, vdi)
+	if errRes != nil {
+		return nil, sdk.ErrInternal(sdk.AppendMsgToErr("could not marshal result to JSON", errRes.Error()))
+	}
+	return res, nil
+}
+
+// params for query 'custom/distr/rewards'
+type QueryRewardsParams struct {
+	Address sdk.AccAddress `json:"address"`
+}
+
+// creates a new instance of QueryRewardsParams
+func NewQueryRewardsParams(address sdk.AccAddress) QueryRewardsParams {
+	return QueryRewardsParams{
+		Address: address,
+	}
+}
+
+type Rewards struct {
+	Total       sdk.Coins           `json:"total"`
+	Delegations []DelegationsReward `json:"delegations"`
+	Commission  sdk.Coins           `json:"commission"`
+}
+
+type DelegationsReward struct {
+	Validator sdk.ValAddress `json:"validator"`
+	Reward    sdk.Coins      `json:"reward"`
+}
+
+func queryRewards(ctx sdk.Context, _ []string, req abci.RequestQuery, k Keeper) ([]byte, sdk.Error) {
+	totalWithdraw := types.DecCoins{}
+	rewards := Rewards{}
+
+	var params QueryRewardsParams
+	err := k.cdc.UnmarshalJSON(req.Data, &params)
+	if err != nil {
+		return nil, sdk.ErrUnknownRequest(sdk.AppendMsgToErr("incorrectly formatted request data", err.Error()))
+	}
+
+	// cache-wrap context as to not persist state changes during querying
+	ctx, _ = ctx.CacheContext()
+	var selfVdi types.ValidatorDistInfo
+	selfValidator := k.stakeKeeper.Validator(ctx, sdk.ValAddress(params.Address))
+	if selfValidator != nil && selfValidator.GetOperator().Equals(sdk.ValAddress(params.Address)) {
+		selfVdi = k.GetValidatorDistInfo(ctx, selfValidator.GetOperator())
+	}
+
+	feePool := k.GetFeePool(ctx)
+
+	// get all delegator rewards
+	operationAtDelegation := func(_ int64, del sdk.Delegation) (stop bool) {
+		validator := k.stakeKeeper.Validator(ctx, del.GetValidatorAddr())
+		vdi := k.GetValidatorDistInfo(ctx, del.GetValidatorAddr())
+		wc := k.GetWithdrawContext(ctx, del.GetValidatorAddr())
+		wc.FeePool = feePool
+		distInfo := k.GetDelegationDistInfo(ctx, del.GetDelegatorAddr(), del.GetValidatorAddr())
+		_, vdi, newFeePool, diWithdraw := distInfo.WithdrawRewards(log.NewNopLogger(), wc, vdi, validator.GetDelegatorShares(), del.GetShares())
+		totalWithdraw = totalWithdraw.Plus(diWithdraw)
+		rewardTruncate, _ := diWithdraw.TruncateDecimal()
+		rewards.Delegations = append(rewards.Delegations, DelegationsReward{del.GetValidatorAddr(), rewardTruncate})
+		if vdi.OperatorAddr.Equals(selfValidator.GetOperator()) {
+			selfVdi = vdi
+		}
+		feePool = newFeePool
+		return false
+	}
+	k.stakeKeeper.IterateDelegations(ctx, params.Address, operationAtDelegation)
+
+	// get all validator rewards
+	if selfVdi.OperatorAddr.Equals(sdk.ValAddress(params.Address)) {
```

Comment:
If the delegator is not the operator of any validators, selfVdi may be a empty struct. `selfVdi.OperatorAddr` may be nil. Shall we judge whether `selfVdi.OperatorAddr` is empty or not?

Reply:
The `address. Equals` function Has checked if it is empty.

Reply:
But `selfVdi` may be nil, should add check `selfVdi == nil` .